File: benchmarks/Java/src/richards/TaskState.java
/*
 * This benchmark is derived from Mario Wolczko's Java and Smalltalk version of
 * Richards.
 * 
 * It is modified based on the SOM version and to use Java 8 features.
 * License details:
 *   http://web.archive.org/web/20050825101121/http://www.sunlabs.com/people/mario/java_benchmarking/index.html
 */
package richards;


class TaskState extends RBObject {
  private boolean packetPending;
  private boolean taskWaiting;
  private boolean taskHolding;

  public boolean isPacketPending() { return packetPending; }
  public boolean isTaskHolding()   { return taskHolding;   }
  public boolean isTaskWaiting()   { return taskWaiting;   }

  public void setTaskHolding(final boolean b) { taskHolding = b; }
  public void setTaskWaiting(final boolean b) { taskWaiting = b; }
  public void setPacketPending(final boolean b) { packetPending = b; }

  public void packetPending() {
    packetPending = true;
    taskWaiting   = false;
    taskHolding   = false;
  }

  public void running() {
    packetPending = taskWaiting = taskHolding = false;
  }

  public void waiting() {
    packetPending = taskHolding = false;
    taskWaiting = true;
  }

  public void waitingWithPacket() {
    taskHolding = false;
    taskWaiting = packetPending = true;
  }

  public boolean isRunning() {
    return !packetPending && !taskWaiting && !taskHolding;
  }

  public boolean isTaskHoldingOrWaiting() {
    return taskHolding || (!packetPending && taskWaiting);
  }

  public boolean isWaiting() {
    return !packetPending && taskWaiting && !taskHolding;
  }

  public boolean isWaitingWithPacket() {
    return packetPending && taskWaiting && !taskHolding;
  }

  public static TaskState createPacketPending() {
    TaskState t = new TaskState();
    t.packetPending();
    return t;
  }

  public static TaskState createRunning() {
    TaskState t = new TaskState();
    t.running();
    return t;
  }

  public static TaskState createWaiting() {
    TaskState t = new TaskState();
    t.waiting();
    return t;
  }

  public static TaskState createWaitingWithPacket() {
    TaskState t = new TaskState();
    t.waitingWithPacket();
    return t;
  }
}

File: benchmarks/Csharp/Benchmarks/Richards/TaskState.cs
namespace AreWeFastYet;

class TaskState : RBObject
{
    public virtual bool IsPacketPending { get; set; }
    public virtual bool IsTaskHolding { get; set; }
    public virtual bool IsTaskWaiting { get; set; }

    public virtual void SetPacketPending()
    {
        IsPacketPending = true;
        IsTaskWaiting = false;
        IsTaskHolding = false;
    }

    public virtual void SetRunning()
    {
        IsPacketPending = IsTaskWaiting = IsTaskHolding = false;
    }

    public virtual void SetWaiting()
    {
        IsPacketPending = IsTaskHolding = false;
        IsTaskWaiting = true;
    }

    public virtual void SetWaitingWithPacket()
    {
        IsTaskHolding = false;
        IsTaskWaiting = IsPacketPending = true;
    }

    public virtual bool IsRunning
        => !IsPacketPending && !IsTaskWaiting && !IsTaskHolding;

    public virtual bool IsTaskHoldingOrWaiting
        => IsTaskHolding || (!IsPacketPending && IsTaskWaiting);

    public virtual bool IsWaiting
        => !IsPacketPending && IsTaskWaiting && !IsTaskHolding;

    public virtual bool IsWaitingWithPacket
        => IsPacketPending && IsTaskWaiting && !IsTaskHolding;

    public static TaskState CreatePacketPending()
    {
        var t = new TaskState();
        t.SetPacketPending();
        return t;
    }

    public static TaskState CreateRunning()
    {
        var t = new TaskState();
        t.SetRunning();
        return t;
    }

    public static TaskState CreateWaiting()
    {
        var t = new TaskState();
        t.SetWaiting();
        return t;
    }

    public static TaskState CreateWaitingWithPacket()
    {
        var t = new TaskState();
        t.SetWaitingWithPacket();
        return t;
    }
}
